From 4ce57aded9ed687c333a02018fd8d00f2c9a2d71 Mon Sep 17 00:00:00 2001
From: Laurentiu Palcu <laurentiu.palcu@nxp.com>
Date: Wed, 21 Feb 2018 10:00:55 +0200
Subject: [PATCH 3370/5242] MLK-17459-3: drm: imx: dcss: fixes for compressed
 format cropping

commit  6c8af4accc5ac452d26ceab6dd264acde9ef336b from
https://source.codeaurora.org/external/imx/linux-imx.git

Cropping of compressed formats seems problematic and we cannot up-align
in this case. For compressed formats we need to down-align both the
width and height.

Signed-off-by: Laurentiu Palcu <laurentiu.palcu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/dcss/dcss-plane.c |   11 ++++++++++-
 drivers/gpu/imx/dcss/dcss-dtrc.c      |   13 ++++++++++---
 2 files changed, 20 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/imx/dcss/dcss-plane.c b/drivers/gpu/drm/imx/dcss/dcss-plane.c
index 42db8fd..8786ca9 100644
--- a/drivers/gpu/drm/imx/dcss/dcss-plane.c
+++ b/drivers/gpu/drm/imx/dcss/dcss-plane.c
@@ -359,6 +359,7 @@ static void dcss_plane_atomic_update(struct drm_plane *plane,
 	bool modifiers_present = !!(fb->flags & DRM_MODE_FB_MODIFIERS);
 	u32 src_w, src_h, adj_w, adj_h;
 	struct drm_rect disp, crtc, src, old_src;
+	u32 scaler_w, scaler_h;
 
 	if (!state->fb)
 		return;
@@ -426,8 +427,16 @@ static void dcss_plane_atomic_update(struct drm_plane *plane,
 			 src_w, src_h, adj_w, adj_h);
 	dcss_plane_atomic_set_base(dcss_plane);
 
+	if (fb->modifier == DRM_FORMAT_MOD_VSI_G2_TILED_COMPRESSED) {
+		scaler_w = src.x1 ? adj_w : src_w;
+		scaler_h = src.y1 ? adj_h : src_h;
+	} else {
+		scaler_w = src_w;
+		scaler_h = src_h;
+	}
+
 	dcss_scaler_setup(dcss_plane->dcss, dcss_plane->ch_num,
-			  pixel_format, src_w, src_h,
+			  pixel_format, scaler_w, scaler_h,
 			  crtc.x2 - crtc.x1,
 			  crtc.y2 - crtc.y1,
 			  drm_mode_vrefresh(&crtc_state->mode));
diff --git a/drivers/gpu/imx/dcss/dcss-dtrc.c b/drivers/gpu/imx/dcss/dcss-dtrc.c
index dee4f75..37bf782 100644
--- a/drivers/gpu/imx/dcss/dcss-dtrc.c
+++ b/drivers/gpu/imx/dcss/dcss-dtrc.c
@@ -363,9 +363,16 @@ void dcss_dtrc_set_res(struct dcss_soc *dcss, int ch_num, struct drm_rect *src,
 		goto exit;
 	}
 
-	/* align the image size */
-	xres = (xres + 0x7f) & ~0x7f;
-	yres = (yres + 0xf) & ~0xf;
+	/* align the image size: down align for compressed formats */
+	if (ch->format_modifier == DRM_FORMAT_MOD_VSI_G2_TILED_COMPRESSED && src->x1)
+		xres = xres & ~0x7f;
+	else
+		xres = (xres - 1 + 0x7f) & ~0x7f;
+
+	if (ch->format_modifier == DRM_FORMAT_MOD_VSI_G2_TILED_COMPRESSED && src->y1)
+		yres = yres & ~0xf;
+	else
+		yres = (yres - 1 + 0xf) & ~0xf;
 
 	src->x1 &= ~1;
 	src->x2 &= ~1;
-- 
1.7.9.5

